### PR TITLE
Add metadata

### PR DESCRIPTION
**Pull request by ChrisHuynh333**
## What does this PR do and why?
This PR allows users to add metadata to a sample through the website UI.

## Screenshots or screen recordings
Add metadata:
image
image

Add metadata with keys that already exist:
image
image

Add both new and existing metadata:
image
image

Add metadata after deleting fields on form:
[Screencast from 2024-02-12 09:37:35 AM.webm](https://github.com/phac-nml/irida-next/assets/82407232/e2ccfc05-4a46-4afe-8694-f7fe535ae7f1)

## How to set up and validate locally
1. Navigate to a sample's metadata page
2. Add metadata and ensure they properly appear on the metadata table
3. Ensure adding metadata with keys that already exist are not added.
5. Fill out multiple form fields and then delete some fields. Ensure the deleted fields are not submitted and added to metadata, and only the remaining fields are added.
6. In the add metadata modal with only one field, fill out the metadata field, and then delete the field. Ensure the field still exists but is once again blank.

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [x] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

### CODE REVIEW

File: app/views/projects/samples/metadata/_form.html.erb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  <div id='test' class="grid gap-2" data-controller="projects--samples--metadata--add">
+    <div data-projects--samples--metadata--add-target="metadataToAdd"></div>
+    <div id='input-field-0' class='inputField'>
+      <div id="field-label-0" class="text-slate-700 dark:text-slate-400 text-sm mb-2">
+        <%= t(".field")%> 1:
+      </div>
+
+      <div class="flex space-x-2 mb-2">
+        <div class="relative z-0 w-full group">
+          <input type="text" name="key-0" id="key-0" class="block py-2.5 px-0 w-full text-sm text-gray-900 bg-transparent border-0 border-b-2 border-gray-300 appearance-none dark:text-white dark:border-gray-600 dark:focus:border-primary-500 focus:outline-none focus:ring-0 focus:border-primary-500 peer" placeholder=" " required />
```
Comment (joshsadam) [inline app/views/projects/samples/metadata/_form.html.erb:12]: **nitpick:** Please update all the `gray` to `slate`

Comment (ChrisHuynh333) [reply app/views/projects/samples/metadata/_form.html.erb]: Fixed in 24e373a7374b8f112cc0f9b44d43fcfbf1baad80

**Review by joshsadam — CHANGES_REQUESTED** [review]
Can you try to simplify this by using a template, I left an example in one of the comments.

File: app/javascript/controllers/projects/samples/metadata/create_controller.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            <label for="value-${id}" class="peer-focus:font-medium absolute text-sm text-slate-500 dark:text-slate-400 duration-300 transform -translate-y-6 scale-75 top-3 -z-10 origin-[0] peer-focus:start-0 rtl:peer-focus:translate-x-1/4 rtl:peer-focus:left-auto peer-focus:text-primary-500 peer-focus:dark:text-primary-500 peer-placeholder-shown:scale-100 peer-placeholder-shown:translate-y-0 peer-focus:scale-75 peer-focus:-translate-y-6">${this.valueLabel}</label>
+          </div>
+          <button type="button" data-action="projects--samples--metadata--create#removeField" class="ml-auto bg-white text-slate-400 hover:text-slate-900 rounded-lg focus:ring-2 focus:ring-slate-300 p-1.5 hover:bg-slate-100 inline-flex items-center justify-center h-8 w-8 dark:text-slate-500 dark:hover:text-white dark:bg-slate-800 dark:hover:bg-slate-700" aria-label="Remove field">
```
[the remark this answers is not in the document]
Comment (ChrisHuynh333) [reply app/javascript/controllers/projects/samples/metadata/create_controller.js]: Fixed in e331870eced747eb2bfbabdf35cc0f72f11eeec6

File: app/javascript/controllers/projects/samples/metadata/create_controller.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            <span class="h-5 w-5 Viral-Icon pointer-events-none">
+              <svg fill="none" stroke="currentColor" stroke-width="1.5" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg" aria-hidden="true" class="Viral-Icon__Svg icon-x_mark pointer-events-none" focusable="false">
```
[the remark this answers is not in the document]
Comment (ChrisHuynh333) [reply app/javascript/controllers/projects/samples/metadata/create_controller.js]: Fixed in e331870eced747eb2bfbabdf35cc0f72f11eeec6

File: app/javascript/controllers/projects/samples/metadata/create_controller.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChrisHuynh333) [reply app/javascript/controllers/projects/samples/metadata/create_controller.js]: Fixed in e331870eced747eb2bfbabdf35cc0f72f11eeec6 and 79e54f03c4ebd297a464e16f3037781b766b76c4

File: app/javascript/controllers/projects/samples/metadata/create_controller.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+  connect() {
+    // Grabs labels for field manipulation in case of translations
+    this.fieldLabel = this.inputsContainerTarget.querySelector('#field-label').innerText.split(" ")[0]
+    this.keyLabel = this.inputsContainerTarget.querySelector('#key-label').innerText
+    this.valueLabel = this.inputsContainerTarget.querySelector('#value-label').innerText
+  }
+
+  // Add new field with next ID increment
+  addNewField() {
```
[the remark this answers is not in the document]
Comment (ChrisHuynh333) [reply app/javascript/controllers/projects/samples/metadata/create_controller.js]: As discussed, fix start here for your thoughts/review d55243e76791f300a65106c3822a98a4c0ca523e

File: app/javascript/controllers/projects/samples/metadata/create_controller.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
+import { Controller } from "@hotwired/stimulus";
+
+export default class extends Controller {
+  static targets = ["metadataToAdd", "fieldsContainer", "fieldTemplate"]
+
+  connect() {
+    this.addField()
```
Comment (joshsadam) [inline app/javascript/controllers/projects/samples/metadata/create_controller.js:7]: **nitpick:** When a new field set is added can you autofocus on the key input please.

File: app/assets/stylesheets/application.tailwind.css [hunk 6]
```diff
@@ -299,9 +299,7 @@ body {
   }
```
[the remark this answers is not in the document]
Comment (ChrisHuynh333) [reply app/assets/stylesheets/application.tailwind.css:299]: Yes this looks like one of my extensions fixing the formatting of this page. I can undo if you'd like.

File: app/javascript/controllers/projects/samples/metadata/create_controller.js (hunk 5 above)
Comment (ChrisHuynh333) [reply app/javascript/controllers/projects/samples/metadata/create_controller.js:7]: Added in 29a8aed5011098652a3ad377856a8f9bcdcffd8a

File: app/assets/stylesheets/application.tailwind.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (joshsadam) [reply app/assets/stylesheets/application.tailwind.css:299]: Please if possible

File: app/javascript/controllers/projects/samples/metadata/create_controller.js (hunk 5 above)
Comment (joshsadam) [reply app/javascript/controllers/projects/samples/metadata/create_controller.js:7]: Awesome, thanks.

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
This is looking great just have one comment below.

File: app/assets/stylesheets/application.tailwind.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (ChrisHuynh333) [reply app/assets/stylesheets/application.tailwind.css:298]: Fixed in 8c5cd96cb1883ea1f06055b93ef49ab6affbb5e3

File: config/locales/en.yml [hunk 7]
```diff
@@ -979,13 +991,15 @@ en:
       metadata:
         sample_does_not_belong_to_project: "Sample '%{sample_name}'' does not belong to project '%{project_name}'"
         empty_metadata: "No metadata was received to update sample '%{sample_name}'"
+        fields:
+          single_all_keys_exist: Metadata key '%{key}' already exists.
+          multi_all_keys_exist: Metadata keys '%{keys}' already exists.
```
[the remark this answers is not in the document]
Comment (ChrisHuynh333) [reply config/locales/en.yml]: Can you tell I didn't like english class :D? 8c5cd96cb1883ea1f06055b93ef49ab6affbb5e3

File: app/assets/stylesheets/application.tailwind.css (hunk 6 above)
[the remark this answers is not in the document]
Comment (joshsadam) [reply app/assets/stylesheets/application.tailwind.css:298]: Works for me.  Formatting of that file can be a pain to keep in sync.  We might have to find a consistent formatter.

**Review by joshsadam — APPROVED** [review]
This looks great, thanks for the changes

File: config/locales/en.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply config/locales/en.yml]: Haha I didn't either!

**Review by deepsidhu85 — APPROVED** [review]

**Pull request merged**